platform/pulp/platform/models/repository.py

```diff
@@ -1,10 +1,268 @@
+
+    class Meta:
+        abstract = True
+
+
+class RepositoryPlugin(Plugin):
+    """
+    An abstract plugin associated with a repository.
+
+    Fields:
+
+    Relations:
+
+    :cvar repository: The associated repository.
+    :type repository: models.ForeignKey
+    """
+
+    repository = models.ForeignKey(Repository, on_delete=models.CASCADE)
+
+    class Meta:
+        abstract = True
+        unique_together = (
+            'name',
+            'type',
+            'repository'
+        )
+
+
+class GroupPlugin(Plugin):
+    """
+    An abstract plugin associated with a repository group.
+
+    Fields:
+
+    Relations:
+
+    :cvar group: The associated repository group.
+    :type group: models.ForeignKey
+    """
+
+    group = models.ForeignKey(RepositoryGroup, on_delete=models.CASCADE)
+
+    class Meta:
+        abstract = True
+        unique_together = (
+            'name',
+            'type',
+            'group'
+        )
+
+
+class Importer(RepositoryPlugin):
+    """
+    A content importer associated with a repository..
+
+    Fields:
+
+    :cvar scratchpad: Arbitrary information stashed by the importer.
+    :type scratchpad: fields.GenericRelation
+
+    :cvar last_sync: When the last successful synchronization occurred.
+    :type last_sync: fields.GenericRelation
+
+    Relations:
+
+    """
+    last_sync = models.DateTimeField(blank=True, null=True)
+
+    scratchpad = fields.GenericRelation(Scratchpad)
+
+
+class Distributor(RepositoryPlugin):
+    """
+    A content distributor.
+
+    Fields:
+
+    :cvar auto_publish: Indicates that the distributor may publish automatically
+        when the associated repository's content has changed.
+    :type auto_publish: fields.BooleanField
+
+    :cvar last_publish: When the last successful publish occurred.
+    :type last_publish: fields.GenericRelation
+
+    Relations:
+
+    """
+    auto_publish = models.BooleanField(default=False)
+    last_publish = models.DateTimeField(blank=True, null=True)
+
+
+class GroupDistributor(GroupPlugin):
+    """
+    A group content distributor.
+
+    Fields:
+
+    :cvar last_publish: When the last successful publish occurred.
+    :type last_publish: fields.GenericRelation
+
+    Relations:
+
+    """
+    last_publish = models.DateTimeField(blank=True, null=True)
+
+
+class RepositoryContent(Model):
+    """
+    Association between a repository and its contained content.
+
+    Fields:
+
+    :cvar created: When the association was created.
+    :type created: fields.DateTimeField
+
+    Relations:
+
+    :cvar content: The associated content.
+    :type content: models.ForeignKey
+
+    :cvar repository: The associated repository.
+    :type repository: models.ForeignKey
+    """
+    created = models.DateTimeField(auto_now_add=True)
+
+    content = models.ForeignKey('Content', on_delete=models.CASCADE)
+    repository = models.ForeignKey(Repository, on_delete=models.CASCADE)
+
+    def save(self, *args, **kwargs):
```

Comment:
The explicit *through=* (join table) causes methods such as`content.add()` to raise an exception.  Those methods cannot be used.